explainer.md

```diff
@@ -161,119 +160,134 @@ user agent.
 
 ## Usage examples
 
-The simplest possible usage of the API is a single call to `Badge.set` from a
-foreground context (which might be used to show an unread count in an email
-app):
+The API is divided into two parts: one for setting/clearing a badge on the
+current document, and one for setting/clearing a badge applied to a particular
+scope.
+
+To simply set a numeric badge on the current document:
 
 ```js
-Badge.set(getUnreadCount());
+Badge.setForDocument(getUnreadCount());
```

Comment:
I wonder if we can can imply "document" if `{scope}` is missing?:

```
// document
Badge.set(5);
// some explicit scope
Badge.set(5, {scope: ""});
```

Reply:
> What about browsers where setting a document Badge is not supported?

That's something I hope we can avoid. At a minimum, document level badging should be supported, with scope level badging being optional... 

> On a separate note, I'm not convinced that document badges are the most natural default. Most of the time, a developer's intention is probably to badge all pages in an app .

Heh, I take the opposite view :) **P**-WAs are, after all, "Progressive" enhancements... "installed app" is never a default state of an application. In the rare case that an application is promoted to "installed", then that's great for the app - in that `{scope: "path"}` would work.

Reply:
> Heh, I take the opposite view :) P-WAs are, after all, "Progressive" enhancements... "installed app" is never a default state of an application. In the rare case that an application is promoted to "installed", then that's great for the app - in that {scope: "path"} would work.

I don't mean for an installed application  ("app" is super overloaded term, I probably shouldn't've used it). I mean that conceptually, it is 'Facebook' that has an unread count, not each individual facebook.com tab. Even without installable PWAs there's a 'Facebook' thing that has a notification count (and I don't think that's a single open tab).

Reply:
ok yeah, that's the bit that's still open for debate (and yes, let's stop using "app":)). I'm still wanting sites/developers to self manage their own scopes (through broadcast channel or whatever) - but can understand why the browser managing it takes some of that pain away.

Reply:
> That's something I hope we can avoid. At a minimum, document level badging should be supported, with scope level badging being optional...

The point of this PR (and the direction we're trying to take the API with) is that, after a lot of discussion with many stakeholders (including the TAG review) we realised that these are just two separate APIs with very different properties (one is global, the other is document-local; one has a lifetime beyond that of the document, the other's lifetime is tied to the document; one works from service workers, the other is meaningless in that context).

Given that we have basically identified two different APIs, I don't think user agents need to support one of these APIs before they support the other. We don't initially intend to implement the document-level badging API because it doesn't fit with our priorities, as you don't initially intend to implement the scope-level API. I think these things can exist independently.

Reply:
Apologies, I hadn't seen any updates to https://github.com/w3ctag/design-reviews/issues/387 ? Was that discussion happening elsewhere (particularly around splitting the two APIs)? 

I can definitely live with that (as it's good during this incubation phase!), but it does feel a bit odd/confusing that we will have two APIs that basically do very similar things.

Reply:
Sorry, maybe you were out of that loop.

It's something we discussed offline with @alice (from TAG) and then as a result of that, we added [this section](https://github.com/WICG/badging/blob/master/explainer.md#a-case-for-separation) to the current explainer. (I somehow thought you were part of this discussion because I thought you were pushing for separate APIs from earlier on when I was trying to combine them.)

This PR is just me taking the "case for separation" part of that explainer and updating the rest of the document to do what it suggested.

The main motivation for this was a) the fact that a combined API had a complicated feature detection story (as discussed in the recent comments on the TAG review) and b) the complexity of dealing with nested scopes applied to pages --- we still have nested scopes in the new proposal, but they only apply to "handles" so you are much less likely to have a complex nesting of badge scopes since if you just want to badge individual pages you'll just use the document API.

> I can definitely live with that (as it's good during this incubation phase!), but it does feel a bit odd/confusing that we will have two APIs that basically do very similar things.

It does a bit, but we'll spec them together in one document (they won't be "separate APIs" but just separate methods), and I think we have a good story for what the differences are. (If we let scopes apply to pages and the document API just overrides that, then it becomes murkier and harder to reason about, which is why I favour a clean separation where the scope API does not badge active documents at all, just inactive handles like app icons.)

Reply:
@marcoscaceres Do you have any final thoughts (veto?) about this before I merge?

Reply:
Just going through the document now. However, I'm still wondering if the TAG as a whole will take a position on this, as opposed to individual TAG members? for instance, I've also be talking to @dbaron about `scope` and concerns about complexity (but don't preclude David reaching a different position to mine) -  hence asking for a TAG position that would close  w3ctag/design-reviews#387.